### PR TITLE
this runs through the background and removes any experiments that are…

### PR DESCRIPTION
**Pull request by thomaszurkan-optimizely**
… not in a fresh datafile

### CODE REVIEW

**Review by msohailhussain — COMMENTED** [review]
PR looks good to me. 
Can we add one unit test. 
Load one datafile, set userprofile, then load another datafile which doesn't have the same experiment mapped in userprofile. and call cleanup service and check what result is returned.

**Review by mikeproeng37 — CHANGES_REQUESTED** [review]
Same comment as @msohailhussain , please add unit tests for these methods.

File: OptimizelySDKUserProfileService/OptimizelySDKUserProfileService/OPTLYUserProfileService.m [hunk 1]
```diff
@@ -156,6 +156,26 @@ - (void)removeUserExperimentRecordsForUserId:(nonnull NSString *)userId {
 - (void)removeAllUserExperimentRecords {
     [self.dataStore removeAllUserData];
 }
+
+- (void)removeInvalidExperimentsForAllUsers:(NSArray<NSString *> *)validExperimentIds {
+    
+    NSMutableDictionary*userProflieService = [[self.dataStore getUserDataForType:OPTLYDataStoreDataTypeUserProfileService] mutableCopy];
```
Comment (mikeproeng37) [inline OptimizelySDKUserProfileService/OptimizelySDKUserProfileService/OPTLYUserProfileService.m]: nit: typo in `userProflieService`. Also need a space before it

File: OptimizelySDKShared/OptimizelySDKShared/OPTLYManagerBase.m [hunk 2]
```diff
@@ -67,6 +69,29 @@ @interface OPTLYManagerBase()
 
 @implementation OPTLYManagerBase
 
+- (void)cleanUserProfileService:(NSArray<OPTLYExperiment> *)experiments {
+    if (experiments == nil) return;
+    
+    if (_userProfileService != nil && [(NSObject *)_userProfileService respondsToSelector:@selector(removeInvalidExperimentsForAllUsers:)]) {
```
Comment (mikeproeng37) [inline OptimizelySDKShared/OptimizelySDKShared/OPTLYManagerBase.m:75]: This seems dangerous, as in, should we wrap it in a `try/catch`?

File: OptimizelySDKUserProfileService/OptimizelySDKUserProfileService/OPTLYUserProfileService.h [hunk 3]
```diff
@@ -52,4 +52,8 @@ __attribute((deprecated("Use OPTLYManager initWithBuilder method instead.")));
  **/
 - (void)removeAllUserExperimentRecords;
 
+/**
+ * Cleanup and remove experiments that are not in the valid experiment list passed in.
```
Comment (mikeproeng37) [inline OptimizelySDKUserProfileService/OptimizelySDKUserProfileService/OPTLYUserProfileService.h]: nit: `Clean up...`

File: OptimizelySDKShared/OptimizelySDKShared/OPTLYManagerBase.m [hunk 4]
```diff
@@ -67,6 +69,29 @@ @interface OPTLYManagerBase()
 
 @implementation OPTLYManagerBase
 
+- (void)cleanUserProfileService:(NSArray<OPTLYExperiment> *)experiments {
```
Comment (mikeproeng37) [inline OptimizelySDKShared/OptimizelySDKShared/OPTLYManagerBase.m:72]: Can you also add some unit tests for this method, particularly around failure scenarios

**Review by mikeproeng37 — CHANGES_REQUESTED** [review]
Mostly ok, just need a few more tests

File: OptimizelySDKShared/OptimizelySDKShared/OPTLYManagerBase.m (hunk 2 above)
Comment (thomaszurkan-optimizely) [reply OptimizelySDKShared/OptimizelySDKShared/OPTLYManagerBase.m:75]: respondsToSelector is a test.  I can't think of a possible reason an exception could be thrown here.

File: OptimizelySDKShared/OptimizelySDKShared/OPTLYManagerBase.m (hunk 4 above)
Comment (thomaszurkan-optimizely) [reply OptimizelySDKShared/OptimizelySDKShared/OPTLYManagerBase.m:72]: @mikeng13 done

**Review by mikeproeng37 — APPROVED** [review]
lgtm, just a few small nits

**Pull request merged**